### DISCUSSION TITLE
@VariableName in XPath rule replacement  (due to deprecation and removal from PMD 7)

### DISCUSSION THREAD

**Discussion by pnatashap** [pmd/pmd#4860] [Q&A] [answered]
I working on update rules to the new version of PMD and faced with deprecation of FieldDeclaration/@VariableName. I got the message in log 
```
WARNING: Use of deprecated attribute 'FieldDeclaration/@VariableName' by XPath rule ..., please use VariableDeclaratorId/@Name instead
```

I have previous rule with 
```xml
<rule name="AvoidDirectAccessToStaticFields" message="Static fields should be accessed in a static way [CLASS_NAME.FIELD_NAME]." language="java" class="net.sourceforge.pmd.lang.rule.XPathRule">
    <description>
      Avoid accessing static fields directly.
    </description>
    <priority>3</priority>
    <properties>
      <property name="xpath">
        <value><![CDATA[
          //Name[@Image = //FieldDeclaration[@Static='true']/@VariableName]
        ]]></value>
      </property>
    </properties>
    <example>
  <![CDATA[
  public final class DirectAccessToStaticFields {
    private static int num = 1;

    public static int number() {
        return num; // violation
    }
   public static int correct() {
        return DirectAccessToStaticFields.num; // correct
    }
  }
  ]]>
    </example>
  </rule>
```

And it works fine (but with warning), but I can not rewrite this XPath to work without warnings using the message (or any other way I found). 
Can someone help with it?

**Comment by oowekyala** [comment]
You can try this (didn't test it):
```
//VariableAccess[@Name = //FieldDeclaration[@Static = true()]/VariableDeclarator/VariableDeclaratorId/@Name]
```
Keep in mind `VariableDeclaratorId` has been renamed `VariableId` in the actual PMD 7 release, which ships later this month.

**Comment by pnatashap** [comment]
@oowekyala unfortunately this variant doesn't work :(

**Reply by oowekyala** [reply]
It works on master (modulo the change to VariableId). What version of PMD are you using?

**Comment by adangel** [comment, marked as answer]
Ok, here is the updated rule for various PMD versions:

PMD 6.55.0 - XPath 1.0:

```
//Name[@Image = //FieldDeclaration[@Static='true']/VariableDeclarator/VariableDeclaratorId/@Name]
```
Note: Replaced `@VariableName` with `VariableDeclarator/VariableDeclaratorId/@Name`

PMD 6.55.0 - XPath 2.0:

```
//Name[@Image = //FieldDeclaration[@Static=true()]/VariableDeclarator/VariableDeclaratorId/@Name]
```
Note: The static attribute is now a real boolean value and not a string anymore.
Note 2: You also need to set `<property name="version" value="2.0"/>`.
Note 3: This is the prerequisite for update to PMD 7

PMD 7.0.0-rc4:

```
//VariableAccess[@Name = //FieldDeclaration[@Static=true()]/VariableDeclarator/VariableDeclaratorId/@Name]
```
Note: You don't need to set `<property name="version" value="2.0"/>` anymore - PMD 7 uses XPath 3.1 by default and that's the only supported XPath version.
Note 2: AST changed: The node is now called VariableAccess

PMD 7.0.0:

```
//VariableAccess[@Name = //FieldDeclaration[@Static=true()]/VariableDeclarator/VariableId/@Name]
```
Note: AST changed again: VariableDeclaratorId -> VariableId


I guess, the deprecation notice here https://github.com/pmd/pmd/blob/ef3455348603aa25f86894b9930f05f141f44d20/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTFieldDeclaration.java#L181
Could have been more explicit by saying `replaceWith = "VariableDeclarator/VariableDeclaratorId/@Name"`.


**Tipp**: Use the designer to develop rules: https://docs.pmd-code.org/latest/pmd_userdocs_extending_designer_reference.html

**Reply by pnatashap** [reply]
Thanks! works with xpath 2.0!
